pkg/apis/portworx/v1/portworxdiag.go

```diff
@@ -0,0 +1,101 @@
+package v1
+
+import (
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// PortworxDiagResourceName is name for "portworxdiag" resource.
+	PortworxDiagResourceName = "portworxdiag"
+	// PortworxDiagResourcePlural is plural for "portworxdiag" resource.
+	PortworxDiagResourcePlural = "portworxdiags"
+)
+
+// PortworxDiagSpec is the spec used to define a portworx diag.
+type PortworxDiagSpec struct {
+	// Configuration for diags collection of the main Portworx component.
+	Portworx *PortworxComponent `json:"portworx,omitempty"`
+}
+
+// Specifies the diags collection configuration for the Portworx component.
+type PortworxComponent struct {
+	// Nodes for which the diags need to be collected. If a volume selector is
+	// also specified, then both the selectors will be honored and the selected
+	// nodes will be a union of both selectors.
+	NodeSelector NodeSelector `json:"nodes,omitempty"`
+	// Volumes for which the diags need to be collected.
+	VolumeSelector VolumeSelector `json:"volumes,omitempty"`
+	// Generates the core dump as well when collecting the diags. Could be useful
+	// to analyze the current state of the system.
+	GenerateCore bool `json:"generateCore,omitempty"`
+}
+
+// Allows selecting all the nodes in the entire cluster or a specific set with
+// node IDs or node label selectors. It will take a union of all the combinations
+// and collect diags from those nodes.
+type NodeSelector struct {
+	// Select all nodes in the Portworx cluster. If set to true, other selectors are ignored.
+	All bool `json:"all,omitempty"`
+	// Ids of the nodes to be selected.
+	IDs []string `json:"ids,omitempty"`
+	// Labels of the volumes to be selected.
+	Labels map[string]string `json:"labels,omitempty"`
+}
+
+// Allows selecting one or more volumes for which diags need to be  collected.
+// It will take a union of volume IDs and volume labels and collect diags for
+// all that match. Currently it will find the Portworx nodes that contain the
+// replicas of the selected volumes and wherever the volume is attached or
+// being consumed from.
+type VolumeSelector struct {
```

Comment:
does this collect node diags of the nodes that have the replicas? not just logs related to the volume activity?

Reply:
Yes all nodes that have replicas of the volume and where the volumes is attached. In case of shared volumes even the nodes where the application is running (clients of shared volumes).